src/network/nm/wireless.cpp

```diff
@@ -0,0 +1,383 @@
+		}
+	}
+
+	if (selectedConn && this->mReferenceConn != selectedConn) {
+		this->mReferenceConn = selectedConn;
+	}
+}
+
+void NMWirelessNetwork::setActiveApPath(const QDBusObjectPath& path) {
+	if (this->mActiveApPath == path) return;
+	this->mActiveApPath = path;
+}
+
+void NMWirelessNetwork::setState(NMConnectionState::Enum state) {
+	if (this->mState == state) return;
+	this->mState = state;
+	emit this->stateChanged(state);
+}
+
+void NMWirelessNetwork::setReason(NMConnectionStateReason::Enum reason) {
+	if (this->mReason == reason) return;
+	this->mReason = reason;
+	emit this->reasonChanged(reason);
+};
+
+void NMWirelessNetwork::setKnown(bool known) {
+	if (this->mKnown == known) return;
+	this->mKnown = known;
+	emit this->knownChanged(known);
+}
+
+void NMWirelessNetwork::updateSignalStrength() {
+	quint8 selectedStrength = 0;
+	NMAccessPoint* selectedAp = nullptr;
+
+	for (auto* ap: this->mAccessPoints.values()) {
+		// Always prefer the active AP if found.
+		if (ap->path() == this->mActiveApPath.path()) {
+			selectedStrength = ap->signalStrength();
+			selectedAp = ap;
+			break;
+		}
+
+		// Otherwise, track the strongest signal.
+		if (selectedStrength <= ap->signalStrength()) {
+			selectedStrength = ap->signalStrength();
+			selectedAp = ap;
+		}
+	}
+
+	if (selectedStrength != this->mSignalStrength) {
+		this->mSignalStrength = selectedStrength;
+		emit this->signalStrengthChanged(selectedStrength);
+	}
+
+	if (selectedAp && this->mReferenceAp != selectedAp) {
+		this->mReferenceAp = selectedAp;
+		NMWirelessSecurityType::Enum selectedSecurity = selectedAp->security();
+		if (this->mSecurity != selectedSecurity) {
+			this->mSecurity = selectedSecurity;
+			emit this->securityChanged(selectedSecurity);
+		}
+	}
+}
+
+void NMWirelessNetwork::addAccessPoint(NMAccessPoint* ap) {
+	if (this->mAccessPoints.contains(ap->path())) return;
+	this->mAccessPoints.insert(ap->path(), ap);
+	// clang-format off
+	QObject::connect(ap, &NMAccessPoint::signalStrengthChanged, this, &NMWirelessNetwork::updateSignalStrength);
+	QObject::connect(ap, &NMAccessPoint::disappeared, this, [this, ap] { this->removeAccessPoint(ap); });
+	// clang-format on
+	this->updateSignalStrength();
+};
+
+void NMWirelessNetwork::removeAccessPoint(NMAccessPoint* ap) {
+	auto* found = this->mAccessPoints.take(ap->path());
+	if (found) {
+		if (this->mAccessPoints.isEmpty()) {
+			emit this->disappeared();
+		} else {
+			QObject::disconnect(ap, nullptr, this, nullptr);
+			this->updateSignalStrength();
+		}
+	}
+};
+
+void NMWirelessNetwork::addConnectionSettings(NMConnectionSettings* conn) {
+	if (this->mConnections.contains(conn->path())) return;
+	this->mConnections.insert(conn->path(), conn);
+	// clang-format off
+	QObject::connect(conn, &NMConnectionSettings::securityChanged, this, &NMWirelessNetwork::updateReferenceConnection);
+	QObject::connect(conn, &NMConnectionSettings::disappeared, this, [this, conn]() { this->removeConnectionSettings(conn); });
```

Comment:
In general I prefer to cast the signal sender instead of passing a lambda to connect for cases like this.